### ISSUE TITLE
Track time operations painfully slow.

### ISSUE THREAD

**Issue by alatdneg** [PixarAnimationStudios/OpenTimelineIO#263]
Hi all, 

I've noticed that operations that involve clips mapped to the track time range, are all very slow. You don't notice it on small timelines but as soon as you get a few hundred clips, it really has problems.

The underlying issue (I believe) is the operation that maps a clips time range to it's parent, involves generating the sum of all preceding clips in the track, as the number of clips in the track increases this becomes a major bottleneck.

I can see why it works this way, but it makes otio unusable for larger timelines, the only solution that comes to mind is a method of caching these values so they are not regenerated all the time.

cheers!

**Comment by alatdneg** [comment]
range_of_child_at_index in the track object seems to be the bit that needs caching.

**Comment by alatdneg** [comment]
This is an example of binary chopping the list, in my code with a 700 clip track, this change is makes it 30x faster.
Sorry if the python is a bit manky, but it does the job :)

```
    def my_each_child(
        self,
        search_range=None,
        descended_from_type=composable.Composable
    ):
        children = self._children

        if search_range:
            left = 0
            right = len(self._children) - 1
            mid = (right + left) / 2

            needle = search_range.start_time

            # search for left most clip whose last frame falls in the search_range
            while True:
                if self.range_of_child_at_index(mid).end_time_exclusive() <= needle:
                    left = mid + 1
                else:
                    right = mid - 1

                # no change / break loop
                if mid == (right + left) / 2:
                    break

                # continue looking
                mid = (right + left) / 2

            # check whether we found no clips
            if mid < len(self._children) - 1:
                left_most = mid + 1
                left = left_most
                right = len(self._children) - 1
                mid = (right + left) / 2

                needle = search_range.end_time_exclusive()
                
                while True:
                    if self.range_of_child_at_index(mid).start_time >= needle:
                        right = mid - 1
                    else:
                        left = mid + 1
     
                   # no change / break loop
                    if mid == (right + left) / 2:
                        break
        
                    # continue looking
                    mid = (right + left) / 2

                right_most = min(mid, len(self._children) - 1)

                children = self._children[left_most : right_most + 1]
            else:
                children = []

        for child in children:
            # filter out children who are not descended from the specified type
            if (
                descended_from_type == composable.Composable
                or isinstance(child, descended_from_type)
            ):
                yield child

            # for children that are compositions, recurse into their children
            if hasattr(child, "each_child"):
                for valid_child in (
                    c for c in child.each_child(
                        search_range,
                        descended_from_type
                    )
                ):
                    yield valid_child

```

**Comment by ssteinbach** [comment]
Thanks for catching this!  Just to repro your use case, you have a track that has a large number of clips (700+) on it and you're calling `range_of_child_at_index`?

**Comment by alatdneg** [comment]
Yep, I'll check with the powers that be, then create a PR, i've got a cleaner method that uses a modified version of the python bisect functions.

**Comment by alatdneg** [comment]
```
def _bisect_right(a, x, lo=0, hi=None, key=lambda y: y):
    """Return the index where to insert item x in list a, assuming a is sorted.
    The return value i is such that all e in a[:i] have e <= x, and all e in
    a[i:] have e > x.  So if x already appears in the list, a.insert(x) will
    insert just after the rightmost x already there.
    Optional args lo (default 0) and hi (default len(a)) bound the
    slice of a to be searched.
    """

    if lo < 0:
        raise ValueError('lo must be non-negative')
    if hi is None:
        hi = len(a)
    while lo < hi:
        mid = (lo+hi)//2
        if x < key(a[mid]): hi = mid
        else: lo = mid+1
    return lo

def _bisect_left(a, x, lo=0, hi=None, key=lambda y: y):
    """Return the index where to insert item x in list a, assuming a is sorted.
    The return value i is such that all e in a[:i] have e < x, and all e in
    a[i:] have e >= x.  So if x already appears in the list, a.insert(x) will
    insert just before the leftmost x already there.
    Optional args lo (default 0) and hi (default len(a)) bound the
    slice of a to be searched.
    """

    if lo < 0:
        raise ValueError('lo must be non-negative')
    if hi is None:
        hi = len(a)
    while lo < hi:
        mid = (lo+hi)//2
        if key(a[mid]) < x: lo = mid+1
        else: hi = mid
    return lo


def my_each_child(
    self,
    search_range=None,
    descended_from_type=composable.Composable
):
    children = self._children

    if search_range:
        left = _bisect_left(self._children, search_range.start_time, key=lambda y: y.trimmed_range_in_parent().end_time_inclusive())
        right = _bisect_right(self._children, search_range.end_time_inclusive(), left, key=lambda y: y.trimmed_range_in_parent().start_time )
        children = self._children[left : right]
        # else:
        #     children = []

    for child in children:
        # filter out children who are not descended from the specified type
        if (
            descended_from_type == composable.Composable
            or isinstance(child, descended_from_type)
        ):
            yield child

        # for children that are compositions, recurse into their children
        if hasattr(child, "each_child"):
            for valid_child in (
                c for c in child.each_child(
                    search_range,
                    descended_from_type
                )
            ):
                yield valid_child
```

**Comment by alatdneg** [comment]
Here's a simple test.. 
`#!/bin/env python

import opentimelineio_extensions
import opentimelineio
import time

# for every clip in video track 0, find clips in all 
# other tracks that fall within it's track in/out points
# not an unusual operation

if __name__ == "__main__":
    timeline = opentimelineio.adapters.read_from_file(
        "big_timeline.otio"
    )
    start_time = time.time()
    
    for clip in timeline.video_tracks()[0]:
        search_range = clip.trimmed_range_in_parent()
        sibling_count = 0
        for track in timeline.video_tracks():
            sibling_count += len(list(track.each_clip(search_range)))
    print time.time() - start_time

# on my workstation..
# 295 seconds.

# using my bisect optimisation
# 33 seconds
`

**Comment by alatdneg** [comment]
Not sure how to give you the cleaned timeline... As I can't attach it..

**Comment by ssteinbach** [comment]
GIven that we've landed #404 I think we can close this issue.  Please let us know if its still a problem.
